llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -3165,56 +3181,87 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
   return AbbrevEntryReader(AbbrevData);
 }
 
-Error DIEVisitor::visitDIERef(BinaryStreamReader &DataReader,
-                              unsigned AbbrevIdx, StringRef DIEData,
-                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
-  AbbrevEntryReader AbbrevReader =
-      getAbbrevEntryReader(AbbrevEntries, AbbrevIdx);
+static void
+restoreStack(BinaryStreamReader &Reader, StringRef &Data,
+             std::stack<std::pair<StringRef, unsigned>> &StackOfNodes) {
+  auto DataAndOffset = StackOfNodes.top();
+  Reader = BinaryStreamReader(DataAndOffset.first, llvm::endianness::little);
+  Data = DataAndOffset.first;
+  Reader.setOffset(DataAndOffset.second);
+  StackOfNodes.pop();
+}
 
-  if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
-    StartTagCallback(*MaybeTag, AbbrevIdx);
-  else
-    return MaybeTag.takeError();
+Error DIEVisitor::visitDIERef(unsigned AbbrevIdx,
+                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
 
-  Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
-  if (!MaybeHasChildren)
-    return MaybeHasChildren.takeError();
+  std::stack<std::pair<StringRef, unsigned>> StackOfNodes;
```

Comment:
Have you considered making this a stack of `BinaryStreamReaders`? I *think* this would greatly simplify the code, as I suspect that passing `Data` as a reference to `restoreStack` is not necessary.

Reply:
It would have to be a stack of `BinaryStreamReader &` right?

Reply:
I don't think so (in fact you can't have containers of references), IIRC a BinaryStreamReader is just a pointer + offset + size, so it shouldn't be an issue to copy it around

Reply:
Yeah you need to have a container of ptrs you are right

Reply:
I don't think there is a simple way to get a `StringRef` of the Data that a `BinaryStreamReader` points to

Reply:
I agree that passing Data as a reference to `restoreStack` is weird, I could return it instead, but I don't see a way to return a StringRef of the data a BinaryStreamReader points to

Reply:
I'll write a top level comment on the review, maybe that will help clarify what I mean